File: 2020/day6/day6.py
from os import linesep
from typing import List

with open("input.txt") as f:
    input = f.read()


def part1():
    groups: List[str] = [line.strip().replace('\n', '').replace(' ', '')
                         for line in input.split('\n\n')]

    total = 0
    for group in groups:
        total += len(set([c for c in group]))

    print('Part 1:', total)


def part2():
    groups: List[str] = [line.strip().replace(' ', '')
                         for line in input.split('\n\n')]
    groups: List[List[str]] = [group.split('\n') for group in groups]

    total = 0
    for group in groups:
        s = {}
        for line in group:
            for char in line:
                s[char] = s[char] + 1 if char in s else 1

        for key in s:
            total += 1 if s[key] == len(group) else 0

    print('Part 2:', total)


part1()
part2()

File: 2023/Problems/Day6.cs
using Problems.Attributes;

namespace Problems;

[Day(6)]
public class Day6 : IProblem
{
    [SampleFile("day6.sample.txt", 288)]
    [PuzzleFile("day6.txt", Expected = 393_120)]
    public int Part1(string[] input)
    {
        var times = ParseMultipleRaces(input[0]);
        var distances = ParseMultipleRaces(input[1]);

        var recordsPerRace = new List<int>();
        foreach (var (time, distance) in times.Zip(distances))
        {
            var records = 0;
            for (var holdSeconds = 0; holdSeconds <= time; holdSeconds++)
            {
                var moveTime = time - holdSeconds;
                var distanceCovered = moveTime * holdSeconds;

                if (distanceCovered > distance)
                {
                    records++;
                }
            }

            recordsPerRace.Add(records);
        }

        return recordsPerRace.Aggregate(1, (acc, curr) => acc * curr);
    }

    [SampleFile("day6.sample.txt", 71_503)]
    [PuzzleFile("day6.txt", Expected = 36_872_656)]
    public int Part2(string[] input)
    {
        var time = ParseSingleRace(input[0]);
        var distance = ParseSingleRace(input[1]);

        var lowestSeconds = long.MaxValue;
        var highestSeconds = long.MinValue;

        for (var holdSeconds = 0; holdSeconds <= time; holdSeconds++)
        {
            var moveTime = time - holdSeconds;
            var distanceCovered = moveTime * holdSeconds;

            if (distanceCovered > distance)
            {
                lowestSeconds = holdSeconds;
                break;
            }
        }

        for (var holdSeconds = time; holdSeconds > 0; holdSeconds--)
        {
            var moveTime = time - holdSeconds;
            var distanceCovered = moveTime * holdSeconds;

            if (distanceCovered > distance)
            {
                highestSeconds = holdSeconds;
                break;
            }
        }

        return (int)highestSeconds - (int)lowestSeconds + 1;
    }

    private static int[] ParseMultipleRaces(string line)
    {
        return line.Split(":")[1]
            .Split(" ", StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
            .Select(int.Parse)
            .ToArray();
    }

    private static long ParseSingleRace(string line)
    {
        return long.Parse(line.Split(":")[1].Replace(" ", ""));
    }
}
